File: Stack/2197. Replace Non-Coprime Numbers in Array/solution.py
class Solution:
    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        def lcm(a, b):
            return a*b//gcd(a,b)

        res = [nums[0]]
        n = len(nums)
        for i in range(1, n):
            if gcd(res[-1], nums[i]) == 1:
                res.append(nums[i])
            else:
                res[-1] = lcm(res[-1], nums[i])
        
        nums = res
        res = [nums[-1]]
        n = len(nums)

        allCoprime = True
        for i in range(n-2, -1, -1):
            if gcd(res[-1], nums[i]) == 1:
                res.append(nums[i])
            else:
                allCoprime = False
                res[-1] = lcm(res[-1], nums[i])

        res.reverse()
        
        if allCoprime: return res
        return self.replaceNonCoprimes(res)

File: Stack/2197. Replace Non-Coprime Numbers in Array/solution.cpp
class Solution {
public:
    vector<int> replaceNonCoprimes(vector<int>& nums) {
        vector<int> stk;
        for (auto& num : nums) {
            if (!stk.empty()) {
                long long num2 = stk.back();
                long long GCD = gcd(num2, num);
                while (GCD > 1) {
                    stk.pop_back();
                    long long LCM = num2 * num / GCD;
                    num = LCM;

                    if (!stk.empty()) {
                        num2 = stk.back();
                        GCD = gcd(num2, num);
                    } else {
                        break;
                    }
                }
                stk.push_back(num);
            } else {
                stk.push_back(num);
            }
        }
        return stk;
    }
};